nightingale_perception/config/movo_body_filter.yaml

```diff
@@ -0,0 +1,25 @@
+cloud_filter_chain:
+- name: body_filter
+  type: robot_body_filter/RobotBodyFilterPointCloud2
+  params:
+     #frames/fixed: "odom"
+     frames/sensor: "camera_link" # might be camera_depth_optical_link
+     sensor/point_by_point: False
+     sensor/min_distance: 0.05
+     body_model/inflation/scale: 1.07
+     body_model/inflation/padding: 0.03
+     filter/do_shadow/test: True
```

Comment:
filter/do_shadow_test (bool, default true)
If true, the filter will mark points shadowed by robot body as SHADOW. If false, such points will be marked OUTSIDE.

i dont think i quite understand the shadow test. why would we care about the shadow of the robot? especially since the output of this node is a laser scan which does not segment obstacles, right? is it just for visualization purposes?

It defaults to true so ig we want it?

Reply:
From the tutorial in the constructsim it says that the shadow test is used to deal with noise caused around the arm by light refraction so I thought any points which appear behind the robot but should not be there would be removed. I can turn it off since it apparently is computationally heavy and if it becomes an issue I can turn it back on

Here' s the blurb if you want to read it cause otherwise you'd have to make the constructsim account and stuff:

Task 3: Shadow Filtering
Shadows are basically invalid points in free space that are generated by light refraction on edges of the robot's body. In simulation, there is nothing like refraction, but several settings can lead to these points being generated, too (like additive gaussian noise). Shadow filter is basically a raytracer that casts a ray from each measured point towards the sensor. If this ray goes through the robot's (inflated) body, the ray is considered invalid and is filtered out. Using the shadow filter allows to keep the collision model tight while still removing the shadow points relatively far from the robot body.

Raytracing is a relatively expensive computational operation, so always verify if your filter is realtime-capable if you enable shadow filtering. What helps is ignoring links that physically can't be seen by the lidar and thus cannot generate any shadow points. It also helps to keep the robot model simple and made of basic geometric shapes instead of meshes, or at least decreasing the mesh resolution. Also take into account that robot_body_filter always uses convex hulls of meshes, so if you use non-convex meshes, you have to either subdivide them into several convex parts, or count with the fact that it is the convex hull that is used. There is also a setting for maximum distance of the points considered for shadow filtering, and this can help a lot if a large portion of the pointclouds is far from the sensor (we assume the body-generated shadow points are all located nearby the robot's body).

Reply:
Tested with shadow test False and the performance seems around the same so I'll leave it off